Use the freshest data on mount.

MMMalik opened the pull request:
I've reproduced the described bug via a new test case. The issue was fixed by reading `data`  after `editor` initialization.

The bug can be also reproduced by slightly altering `EditorTypes` component inside `samples/main.js` file:

```
const [ data, setData ] = useState( startingData );
const [ readOnly, setReadOnly ] = useState( false );

React.useEffect( () => {
	setData( 'Test' );
}, [] );

// pass `data` as a prop
...
```

Please note that I haven't moved the whole props destructuring to `then` callback as some of the props must be read beforehand.

Closes #114.
Closes #127.

MMMalik commented:
On a sidenote, I have committed with my personal email. It's now fixed for my future commits but if you think it's necessary I can re-open this PR with correct email configured.

MMMalik commented:
I noticed one more related issue while investigating [this issue](https://github.com/ckeditor/ckeditor4-react/issues/127). 

1. Alter `EditorTypes` with the following snippet:

```
const [ data, setData ] = useState( startingData );
const [ readOnly, setReadOnly ] = useState( false );

React.useEffect( () => {
	setTimeout( function() {
		setData( 'Test' );
	}, 0 );
}, [] );

// pass `data` as a prop
...
```

2. Then open `http://localhost:8080/samples/#/events`, then navigate to `http://localhost:8080/samples/#/`.

3. Observe that `Classic editor` is still populated with `startingData`.

It looks like there are other racing conditions to consider. I will investigate it further.

f1ames commented:
> On a sidenote, I have committed with my personal email. It's now fixed for my future commits but if you think it's necessary I can re-open this PR with correct email configured.

It's fine :+1: 

As for [https://github.com/ckeditor/ckeditor4-react/pull/187#issuecomment-812582377,](https://github.com/ckeditor/ckeditor4-react/pull/187#issuecomment-812582377,) take your time to check this and if it can be easily fixed please do. If it turns out rather complex and not really related to initial issue it will be good to create a separate PR to not block merging this one.

MMMalik commented:
After closer investigation of [issue 114](https://github.com/ckeditor/ckeditor4-react/issues/114) and [issue 127](https://github.com/ckeditor/ckeditor4-react/issues/127), especially [this comment](https://github.com/ckeditor/ckeditor4-react/issues/127#issuecomment-712809954), I was able to reproduce following bugs:

1. Once the editor is initialized via `componentDidMount`, the initial call to `setData` uses stale value of `this.props.data`. Reading value of `data` prop right before calling `setData` solves this issue.
2. In some scenarios updates of `data` prop might occur before editor is able to process `setData` calls. If this happens, `data` update is lost. One real-life scenario where this problem might occur is setting initial data after receiving response from an api. 

I've added two new test cases to cover two scenarios:

1. Component is mounted, editor is not `loaded` yet, new value of `data` prop is passed.
2. Component is mounted, editor is `loaded` but not it's not yet in `ready` state, new value of `data` prop is passed.

Also, I've added new sample `EditorDataPropChanges` component. It is designed as a showcase to a scenario in which a consumer of the component might use response from a remote api to set initial data of the editor.

sculpt0r commented on src/ckeditor.jsx:
```diff
@@ -84,15 +84,20 @@ class CKEditor extends React.Component {
 				}, null, null, -1 );
 			}
 
+			// Make sure that the freshest data is read from props when editor is ready (#114, #127).
```
```suggestion
			// (#114, #127)
			// Make sure that the freshest data is read from props when editor is ready.
```

MMMalik replied:
Done!

sculpt0r commented on tests/browser/ckeditor.jsx:
```diff
@@ -278,6 +276,54 @@ describe( 'CKEditor Component', () => {
 			} );
 		} );
+
+		it( 'uses the freshest data - component mounted, editor not loaded', () => {
```
Please, add tickets references
```
// (#114, #127)
```
Like we have here: https://github.com/ckeditor/ckeditor4-react/blob/04938080004fc385e232da53fdf4e130dda70971/tests/browser/ckeditor.jsx#L69-L70

[Code style reference](https://github.com/ckeditor/ckeditor4/blob/major/dev/docs/codestyle.md#210-comments)

MMMalik replied:
Done!

sculpt0r commented on tests/browser/ckeditor.jsx:
```diff
@@ -278,6 +276,54 @@ describe( 'CKEditor Component', () => {
+			} );
+		} );
+
+		it( 'uses the freshest data on mount - component mounted, editor loaded', () => {
```
Please, add tickets references
```
// (#114, #127)
```
Like we have here: https://github.com/ckeditor/ckeditor4-react/blob/04938080004fc385e232da53fdf4e130dda70971/tests/browser/ckeditor.jsx#L69-L70

MMMalik replied:
Done!

sculpt0r commented on tests/browser/ckeditor.jsx:
```diff
@@ -278,6 +276,54 @@ describe( 'CKEditor Component', () => {
+				data: initialData
+			} );
+
+			// Component is already mounted, editor is not loaded, `data` prop changes
```
```suggestion
			// Component is already mounted, editor is not loaded, `data` prop changes.
```

MMMalik replied:
Done!

sculpt0r commented on tests/browser/ckeditor.jsx:
```diff
@@ -278,6 +276,54 @@ describe( 'CKEditor Component', () => {
+				data: initialData
+			} );
+
+			// Component is already mounted, editor is not loaded, `data` prop changes
```
```suggestion
			// Component is already mounted, editor is not loaded, `data` prop changes.
```

[Code style reference](https://github.com/ckeditor/ckeditor4/blob/major/dev/docs/codestyle.md#210-comments)

MMMalik replied:
Done!

sculpt0r commented on tests/browser/ckeditor.jsx:
```diff
@@ -278,6 +276,54 @@ describe( 'CKEditor Component', () => {
+			return waitForEditor( component, 'loaded' ).then( editor => {
+				sandbox.spy( editor, 'setData' );
+
+				// Component is already mounted, editor is loaded but not ready, `data` prop changes again
```
```suggestion
				// Component is already mounted, editor is loaded but not ready, `data` prop changes again.
```

MMMalik replied:
Done!

sculpt0r reviewed (changes requested):
Does this PR also close #127? If so, please add another `Closes` section.

Overall it looks good :+1: :smiley:

There are some small style corrections - please read code comments.

MMMalik commented on samples/main.js:
```diff
@@ -78,6 +82,56 @@ function EditorTypes() {
+		</main>
+	);
+}
+
```
Done

MMMalik commented on .gitignore:
```diff
@@ -4,3 +4,4 @@ coverage/
 dist/
 sample/dist/
 react-tests/
+*debug.log
```
I've added this entry to cover `debug.log` file that was generated on my machine when running tests. Maybe it's specific to my configuration but I believe that keeping this pattern doesn't hurt, especially that it covers popular `npm-debug.log` file too.

MMMalik commented on samples/main.js:
```diff
@@ -220,6 +274,12 @@ function debounce( fn, wait ) {
 	};
 }
 
+function timeout( t ) {
```
I've changed the argument name to `time`. However, I'm not sure what would be a better name for promisified `setTimeout`. `timeout` seems to be customary name for this functionality. Alternatively, I can remove it and use nested callbacks instead.

sculpt0r replied:
I was thinking about sth like `sleep`, but indeed, `timeout` matches the most here...

MMMalik commented:
Thanks @sculpt0r for careful review. I've added adjustments as per your suggestions. Regarding #127 issue, then two different issues were reported there. [One of them](https://github.com/ckeditor/ckeditor4-react/issues/127#issuecomment-712809954) is reproducible (it's not the original one though) and I believe that I fixed it here. The original one was not backed with a reproducible test case though. In overall, I believe that we can close #127.

sculpt0r reviewed (approved):
> Thanks @sculpt0r for careful review. I've added adjustments as per your suggestions. Regarding #127 issue, then two different issues were reported there. One of them is reproducible (it's not the original one though) and I believe that I fixed it here. The original one was not backed with a reproducible test case though. In overall, I believe that we can close #127.

Based on [last comment](https://github.com/ckeditor/ckeditor4-react/issues/127#issuecomment-740286163) - it looks like a fix for a common source of bugs. However, you need to write separate `Closes #NUMBER` to bind GH to another issue. It will be visible in `Linked issues`. I corrected this.
![image](https://user-images.githubusercontent.com/6498567/114133317-2b492700-9906-11eb-8965-45d8ceea1c23.png)

So... LGTM :tada: :smiley: :confetti_ball: 
@MMMalik, isn't it your first merged PR? :>
